Stats: Add new TierUpgradeSlider component.

a8ck3n opened the pull request:
Basic component with some CSS to get started with.

<!--
Link a related issue to this PR. If the PR does not immediately resolve the issue,
for example, it requires a separate deployment to production, avoid
using the "fixes" keyword and instead attach the [Status] Fix Inbound label to
the linked issue.
-->

Related to #83898.

## Proposed Changes

Adds a new `TierUpgradeSlider` component that houses a slider (for pricing tiers) and some localizable text explaining the selected tier. Includes some initial styling. Currently behind the `stats/tier-upgrade-slider` feature flag.

1. The existing upgrade flow.
2. The new slider is currently below the flow as it isn't wired up yet.

<img width="494" alt="SCR-20231117-oiuk" src="https://github.com/Automattic/wp-calypso/assets/40267301/3efb1ae0-16d8-4498-8bfc-99a35408eaab">

## Testing Instructions

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

1. Launch the Calypso Live link.
3. Navigate to **Stats → Traffic**.
4. Click the "Upgrade my Stats" button in the upgrade notice.
5. Select the commercial upgrade option.
6. Update the URL to enable the slider: `flags=stats/tier-upgrade-slider` — see example URL below
7. Confirm slider is visible.

OR:

1. Launch the Calypso Live link.
2. Navigate to **Stats → Traffic**.
3. Update URL to: `http://calypso.localhost:3000/stats/purchase/SITE_SLUG?productType=commercial&flags=stats/type-detection,stats/tier-upgrade-slider`
4. Confirm slider is visible.

## Pre-merge Checklist

<!--
Complete applicable items on this checklist **before** merging into trunk. Inapplicable items can be left unchecked.

Both the PR author and reviewer are responsible for ensuring the checklist is completed.
-->

- [ ] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [ ] https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [ ] Have you checked for TypeScript, React or other console errors?
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
- [ ] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-ajp-p2)?

a8ck3n committed:
Add new TierUpgradeSlider component.

Basic component with some CSS to get started with.

grzegorz-cp commented on client/my-sites/stats/stats-purchase/styles.scss:
```diff
@@ -134,6 +134,41 @@ $button-padding: 8px 32px;
+	}
+}
+.stats-tier-upgrade-slider__info-message {
+	color: green;
```
Let's pull the colours from here and the hex values above to variables. It will be easier to update.

grzegorz-cp commented on client/my-sites/stats/stats-purchase/stats-purchase-tier-upgrade-slider.tsx:
```diff
@@ -0,0 +1,92 @@
+	const handleSliderChange = ( value: number ) => {
+		// console.log( 'slider did change. value: ', value );
+		// console.log( 'plan info: ', plans[value] );
+		setCurrentPlanIndex( value );
```
It's ok for now, but if the final source will have IDs, it will be better to find the object based on that rather than the array index.

grzegorz-cp commented on client/my-sites/stats/stats-purchase/stats-purchase-tier-upgrade-slider.tsx:
```diff
@@ -0,0 +1,92 @@
+				</div>
+			</div>
+			<PricingSlider
+				className={ `${ COMPONENT_CLASS_NAME }__slider` }
```
It would be nice to avoid mixing the classes.
I'd prefer to have a local const `COMPONENT_CLASS_NAME` set to `stats-tier-upgrade-slider` and simplify the classes above.

Maybe rather than importing this external const here it would be better to get that class name as a prop.

a8ck3n replied:
I didn't look too closely at this but yes, that sounds like the better approach. Thanks!

grzegorz-cp reviewed (commented):
Great start, I added a few things to consider.

annacmc committed:
add & style marks

a8ck3n committed:
Update types for mark prop.

a8ck3n committed:
Simplify marks setup.

a8ck3n committed:
Remove unused interface definition.

a8ck3n committed:
Remove COMPONENT_CLASS_NAME const.

a8ck3n committed:
Add support for className prop.

a8ck3n committed:
Move our styles to a new file.

a8ck3n committed:
Separate out the data fetch logic for now.

a8ck3n committed:
Translate UI strings.

a8ck3n committed:
Remove comment.

grzegorz-cp commented on client/my-sites/stats/stats-purchase/stats-purchase-tier-upgrade-slider.scss:
```diff
@@ -0,0 +1,51 @@
+
+	.jp-components-pricing-slider__thumb {
+		border-radius: 4px;
+		border-color: #008710;
```
Let's put the colours in variables at the top of the file.

grzegorz-cp commented on client/my-sites/stats/stats-purchase/stats-purchase-tier-upgrade-slider.scss:
```diff
@@ -0,0 +1,51 @@
+		height: 4px;
+	}
+
+	.jp-components-pricing-slider__control .mark {
```
Could you split this into two selectors: `.jp-components-pricing-slider__control { .mark { } }`?

grzegorz-cp commented on client/my-sites/stats/stats-purchase/stats-purchase-tier-upgrade-slider.tsx:
```diff
@@ -0,0 +1,73 @@
+		setCurrentPlanIndex( value );
+	};
+
+	const localizedStrings = getLocalizedStrings();
```
I think it's better to put the object here if you want to use the object. The function doesn't do anything to the strings (apart from putting it into an object), so we don't have to separate it. It's also easier to read when the strings are here rather than checking what's happening in the function.

I'm not sure we need to bundle the strings into an object. It would be more transparent if the translation stayed in h2.

a8ck3n replied:
Added to parent issue as a new task. I feel like this is a matter of style though. For me personally, I prefer to separate the string translation calls from the JSX and they are quite noisy. I don't have strong feelings about wrapping them in a function though.

kangzj replied:
It does look clearer this way; however the problem with use `translate` outside of components is that, if user switches languages or the translation file is loaded after rendering, the component wouldn't reflect the changes if that makes sense. As a result, I would recommend always use hook - `useTranslate` 😃 

Code: 
https://github.com/Automattic/wp-calypso/blob/786367864f24125f0add3197dd14aabb959c4c26/packages/i18n-calypso/src/use-translate.js#L20

kangzj replied:
That said, I think you should still be able to create a hook to translate strings with `useTranslate` like your pattern, for example:

```JavaScript
function useTranslationStrings() {
    const translate = useTranslate();
    // useMemo is optional here, as the translation process is pretty light.
    return useMemo( () => ( { message: translate( 'This is a message' ) } ), [ translate ] );
}
```

PS: I didn't test it. You might need to tweak a bit to make it work 😄

a8ck3n replied:
Thanks for this explanation, Jasper. So in the case of `translate` vs `useTranslate` it's not just a matter of style. I'll note this in the list of follow-up tasks. 👍

grzegorz-cp commented on client/my-sites/stats/stats-purchase/stats-purchase-tier-upgrade-slider.tsx:
```diff
@@ -0,0 +1,73 @@
+};
+
+function TierUpgradeSlider( { className }: TierUpgradeSliderProps ) {
+	const componentClassNames = classNames( 'stats-tier-upgrade-slider', className ? className : '' );
```
I don't think we need the ternary operator. If you pass an undefined value to `classnames` I think it would skip it.

grzegorz-cp commented on client/my-sites/stats/stats-purchase/stats-purchase-tier-upgrade-slider.tsx:
```diff
@@ -0,0 +1,73 @@
+	};
+}
+
+type TierUpgradeSliderProps = {
```
Let's place the type at the top of the file or in a dedicated file.

a8ck3n replied:
Added this on to the task list as well.

grzegorz-cp reviewed (changes requested):
Nice progress separating the code. I left a few comments.

grzegorz-cp commented:
The live preview doesn't load for me. It's stuck on the logo.

annacmc committed:
update colors to vars

annacmc committed:
add basic feature flag for flags=stats/tier-upgrade-slider

annacmc committed:
relocate component

a8ck3n committed:
Minor CSS tweak.

a8ck3n committed:
Remove unneeded ternary operator.

a8ck3n committed:
Add slider to commercial upgrade page.

Seems there are two of these currently. Not sure if this belongs on both so will seek some clarification on that.

a8ck3n committed:
Comments++

a8ck3n committed:
Add slider to single item purchase flow.

Behind a feature flag and below the working flow as this is just for testing of the component.

a8ck3n committed:
Move type definition to top of file.

a8ck3n committed:
Minor CSS tweak.

kangzj commented on client/my-sites/stats/stats-purchase/stats-purchase-commercial.tsx:
```diff
@@ -29,6 +31,9 @@ const CommercialPurchase = ( {
 	from,
 }: CommercialPurchaseProps ) => {
 	const translate = useTranslate();
+	// TODO: Confirm this belongs here.
```
Let's skip the old flow for now

kangzj commented on client/my-sites/stats/stats-purchase/stats-purchase-tier-upgrade-slider-utils.tsx:
```diff
@@ -0,0 +1,90 @@
+];
+
+function getPlanTiers(): StatsPlanTier[] {
+	// TODO: Work out data fetch mechanisim.
```
I believe there is existing mechanism that pulls `/products` API which has the `price_tier_list`

a8ck3n replied:
Added a GH Issue for API integration [here](https://github.com/orgs/Automattic/projects/484/views/18?groupedBy%5BcolumnId%5D=Status&sortedBy%5Bdirection%5D=asc&sortedBy%5BcolumnId%5D=Status).

kangzj commented on packages/components/src/pricing-slider/index.tsx:
```diff
@@ -21,6 +20,7 @@ const PricingSlider: React.FC< PricingSliderProps > = ( {
 	onBeforeChange,
 	onAfterChange,
 	renderThumb,
+	marks, // Add marks to the component props
```
Should we default it to `false` so that it doesn't affect the other place that use the component?

a8ck3n replied:
Yes! The underlying slider handles "undefined" correctly (ie: not showing anything) but I don't think it hurts for us to be explicit in our wrapper.

kangzj reviewed (commented):
I'm happy with it, just need to confirm it doesn't affect the current sliders 😄

grzegorz-cp commented:
Looks good considering future updates. I'll approve after we implement Jasper's suggestions for the default `marks` parameter.

a8ck3n committed:
Remove slider from wizard flow.

a8ck3n committed:
Set default value for marks prop.

a8ck3n commented:
> Looks good considering future updates. I'll approve after we implement Jasper's suggestions for the default `marks` parameter.

I've addressed the last bit of feedback and rebased on the trunk.

kangzj reviewed (approved):
Good job @a8ck3n and @annacmc. LGTM! Let's ship it 🚢

grzegorz-cp reviewed (approved):
Great effort!

a8ci18n commented:
This Pull Request is now available for translation here: https://translate.wordpress.com/deliverables/9723994

Some locales (Hebrew, Japanese) have been temporarily machine-translated due to translator availability. All other translations are usually ready within a few days. Untranslated and machine-translated strings will be sent for translation next Monday and are expected to be completed by the following Friday.

Thank you @a8ck3n for including a screenshot in the description! This is really helpful for our translators.
